org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/Pattern.java

```diff
@@ -148,18 +148,27 @@ protected boolean isApplicable(TypeBinding other, BlockScope scope) {
 			scope.problemReporter().incompatiblePatternType(this, other, patternType);
 			return false;
 		}
+		boolean applicable = true;
 		if (patternType.isBaseType()) {
 			PrimitiveConversionRoute route = Pattern.findPrimitiveConversionRoute(this.resolvedType, this.outerExpressionType, scope);
 			if (!TypeBinding.equalsEquals(other, patternType)
 					&& route == PrimitiveConversionRoute.NO_CONVERSION_ROUTE) {
 				scope.problemReporter().incompatiblePatternType(this, other, patternType);
 				return false;
 			}
-		} else if (!checkCastTypesCompatibility(scope, other, patternType, null, true)) {
-			scope.problemReporter().incompatiblePatternType(this, other, patternType);
-			return false;
+		} else {
+			if (!checkCastTypesCompatibility(scope, patternType, other, null, true)) {
+				if (this.enclosingPattern != null) {
+					scope.problemReporter().incompatiblePatternType(this, other, patternType);
+					return false;
+				}
+				applicable = false;
+			}
```

Comment:
The way this is coded it appears there could be a scenario where the cast is illegal but also unsafe. I doubt that to be valid ever.

If I change `applicable = false;` to be `return applicable = false;` I don't see any test failing. 

If this is a valid observation, it could allow for some simplification of the code by we could get rid of the new variable

Reply:
reworked in https://github.com/eclipse-jdt/eclipse.jdt.core/commit/df50953411e7818a54e426a88611a7642e8c7d1b - do you like it better now?